### DISCUSSION TITLE
Issues with Oracle NoSQL Database Connector Plugin

### DISCUSSION THREAD

**Discussion by otaviojava** [oracle/nosql#19] [Q&A] [answered]
I have encountered several issues while trying to use the Oracle NoSQL Database Connector plugin:

1. **Database Section Visibility**:
    - The plugin does not appear under the "Database" section like other NoSQL databases such as MongoDB, Redis, and Cassandra. This makes it difficult to manage and interact with Oracle NoSQL databases within the IDE.

2. **Connection Setup**:
    - The process for setting up a connection to an Oracle NoSQL database is unclear. The documentation lacks detailed instructions, making it challenging to establish a reliable connection.

3. **Single Database Access**:
    - The plugin seems to support only a single database connection at a time. When I try to connect to a new instance (e.g., local, staging, or production), the previous connection is lost. This limitation significantly hinders the ability to work with multiple environments simultaneously.

4. **Local Connection Issues**:
    - Establishing a local connection could be clearer. Since I don't have a local tenant, the connection process is ambiguous and not well-documented.

**Steps to Reproduce**:
1. Install the Oracle NoSQL Database Connector plugin.
2. Attempt to view the database under the "Database" section.
3. Follow the provided instructions to set up a connection to an Oracle NoSQL database.
4. Try to connect to multiple instances and observe the behavior.
5. Attempt to connect to a local instance without a tenant.

**Expected Behavior**:
- The Oracle NoSQL Database should be visible under the "Database" section for easy management.
- The connection setup process should be straightforward and should have clear and detailed instructions.
- The plugin should support multiple database connections simultaneously without losing the previous ones.
- The documentation should provide clear guidance on connecting to a local instance, even without a tenant.

**Actual Behavior**:
- The database does not appear under the "Database" section.
- The connection setup process is unclear and lacks detailed instructions.
- Only one database connection can be active at a time, and setting up a new connection results in losing the previous one.
- It needs to be clarified how to connect to a local instance without a tenant.

**Comment by dario-vega** [comment]
@otaviojava 

Please let me first clarify my understanding. When you talk about **Database Section Visibility**, you refer to the [Database Tools and SQL plugin in IDEA](https://www.jetbrains.com/help/idea/relational-databases.html). This functionality relies on the **Database Tools and SQL plugin**, which is bundled and enabled in IntelliJ IDEA by default. The Database Tools and SQL plugin is not available in **IntelliJ IDEA Community Edition**. The plugin provides support for all the features that are available in DataGrip, the standalone database management environment for **developers**. DataGrip doesn't support Oracle NoSQL Database; I have created the following ticket: [issue DBE-20736](https://youtrack.jetbrains.com/issue/DBE-20736/Add-Oracle-NoSQL-support). Is my understanding good?(**Q1**)

**Another plugin** is our [Oracle NoSQL database connector for IntelliJ](https://plugins.jetbrains.com/plugin/13052-oracle-nosql-database-connector). You can use the Oracle NoSQL Database Cloud Service IntelliJ plugin to browse tables and run queries against the Oracle NoSQL Database Cloud Service, Oracle NoSQL Database on-premises, or the Oracle NoSQL Cloud Simulator. It is compatible with IntelliJ IDEA (**Ultimate and Community**), Android Studio, Aqua, and MPS. And it is free.

Those plugins are not related.

The connection setup process is documented in two different documents depending on the type of connection (Cloud vs. On-premise). I agree it is **extremely bizarre**, especially for someone starting with Oracle NoSQL—one plugin and two documents to read. (@mikebrey thoughts?)

1. Please check [IntelliJ chapter](https://docs.oracle.com/pls/topic/lookup?ctx=nosql-cloud&id=intellij_plugin) for cloud setup and usage.
2. Please check [Nosql Database Administrator's Guide](https://docs.oracle.com/pls/topic/lookup?ctx=nosql-cloud&id=about_intellij_onprem) for on-premises setup and usage.

To help you understand the documentation when you connect to a local instance, are you using the first or second documentation? (**Q2**) You don't need a tenant to connect to a local instance. For a local instance, you need to read the 2nd link (see screenshot below).

![IntelliJ NoSQL connector](https://github.com/oracle/nosql-examples/assets/66302636/ac8db0e1-e126-4cd9-81c9-0b30fe211545)

Finally, yes, in the next version of the plugin (1.5.0), we have **Multiple connections support for cloud, on-prem, and cloud services**. I will test with the preview version again. The only restriction I saw during my first test is that the connections are related to a project. If you start a new project, you need to reconfigure. I opened a ticket for it, but I wonder if a fix is feasible.

![IntelliJ multiple connections](https://github.com/oracle/nosql-examples/assets/66302636/bc9f7cb5-4c32-477e-9f94-9893c25c615e)

### Action Plan

Can you answer questions (Q1) and (Q2) above? Let me know if it clarifies.

Thanks a lot for your feedback. I'll let you know when the new version, 1.5.0, is available.

**Comment by otaviojava** [comment]
@dario-vega 
I am talking about database support, not SQL, but any database.
As you can see, several databases, including NoSQL ones, such as, MongoDB and Redis.

![Screenshot 2024-05-28 at 18 59 31](https://github.com/oracle/nosql-examples/assets/863011/08b6f9c4-a591-45c3-ab0d-0451d35ccffe)

Another point is the developer experience, I mean, it is not easy to include and set instance as another. Could you make more closer to the default ones?

**Reply by dario-vega** [reply]
> @dario-vega I am talking about database support, not SQL, but any database. As you can see, several databases, including NoSQL ones, such as, MongoDB and Redis.

@otaviojava I don't know if you read the link provided [Database Tools and SQL plugin in IDEA](https://www.jetbrains.com/help/idea/relational-databases.html)

The database management functionality in IntelliJ IDEA is supported by the Database Tools and SQL plugin. The plugin provides support for all the features that are available in DataGrip, the standalone database management environment for developers. With the plugin, you can query, create and manage databases. Databases can work locally, on a server, or in the cloud. The plugin supports MySQL, PostgreSQL, Microsoft SQL Server, SQLite, MariaDB, Oracle, Apache Cassandra, and others.

See the full list of supported vendors in [Create a data source](https://www.jetbrains.com/help/idea/managing-data-sources.html#create_a_data_source).

![image](https://github.com/oracle/nosql-examples/assets/66302636/f685314b-462b-4dee-9ab4-e08aae1735ff)

**Reply by otaviojava** [reply]
Thanks, I saw the link now.

**Comment by dario-vega** [comment]
## Summary

When you talk about **Database Section Visibility**, you refer to the [Database Tools and SQL plugin in IDEA](https://www.jetbrains.com/help/idea/relational-databases.html). 

> _This functionality relies on the Database Tools and SQL plugin, which is bundled and enabled in IntelliJ IDEA by default. The Database Tools and SQL plugin is not available in IntelliJ IDEA Community Edition. The plugin provides support for all the features that are available in DataGrip, the standalone database management environment for developers. With the plugin, you can query, create and manage databases. Databases can work locally, on a server, or in the cloud. The plugin supports MySQL, PostgreSQL, Microsoft SQL Server, SQLite, MariaDB, Oracle, Apache Cassandra, and others (See the full list of supported vendors [here](https://www.jetbrains.com/help/idea/managing-data-sources.html#create_a_data_source).)._ 

DataGrip doesn't support Oracle NoSQL Database, I have created the following ticket: [issue DBE-20736](https://youtrack.jetbrains.com/issue/DBE-20736/Add-Oracle-NoSQL-support).

Version 1.5.0 of our [Oracle NoSQL database connector for IntelliJ](https://plugins.jetbrains.com/plugin/13052-oracle-nosql-database-connector) supports **Multiple connections: cloud, on-prem, and cloud services**. **Note**: It is already available on the [Oracle NoSQL Database Downloads](https://www.oracle.com/database/technologies/nosql-database-server-downloads.html) page and will soon be available in the JetBrains Marketplace. In the meantime, you can use the option "installing Plugin from disk "

**Those plugins are not related.**

You can use the Oracle NoSQL Database Cloud Service IntelliJ plugin to browse tables and run queries against the Oracle NoSQL Database Cloud Service, Oracle NoSQL Database on-premises, or the Oracle NoSQL Cloud Simulator. It is compatible with IntelliJ IDEA (Ultimate and Community), Android Studio, Aqua, and MPS. And it is free.

Thanks a lot for your feedback



I will review your feedback with @mikebrey to see how we can improve the documentation. In the meantime, I can create a video tutorial if you continue to have issues when doing the Local Connection.


@otaviojava, let us know if this clarifies your point and let me know if you need a tutorial video.

**Comment by mikebrey** [comment]
On this list of items:

    The Oracle NoSQL Database should be visible under the "Database" section for easy management.
    The connection setup process should be straightforward and should have clear and detailed instructions.
    The plugin should support multiple database connections simultaneously without losing the previous ones.
    The documentation should provide clear guidance on connecting to a local instance, even without a tenant.

I think the first one is addressed.    On the other 3, we can look at more deeply.

**Reply by otaviojava** [reply]
One point is that we usually use IDE for development. Maybe having a localhost connection as a default may be easier for the developers.

**Comment by dario-vega** [comment]
Hi @otaviojava 

As you can see in this screenshot, the new version allows you to configure multiple connections.

![image](https://github.com/oracle/nosql-examples/assets/66302636/f54a3b95-e654-4bc8-9d5a-608ed8be3827)

Our default is Cloud. As a developer using the container image, is it clear to you that you must choose on-premises? Do you believe that we have to create a new entry in the list box? (see screenshot below)

![image](https://github.com/oracle/nosql-examples/assets/66302636/5351ca99-4073-432d-b88b-36dc2f3fb8bc)

I'll talk with the team to see if we can preconfigure a connection to localhost, thinking of a developer using the container image.

Dario

**Reply by otaviojava** [reply]
@dario-vega 
Could you make the SDK path embedded?
![Screenshot 2024-07-20 at 19 57 04](https://github.com/user-attachments/assets/51b66f18-d674-438e-af6b-f32af8fe51f4)

As a user. I have "infinite" internet access, but I do have limited time.
It would be great if the path be there.

I mean:

1. Go to the plugin store
2. Select the type
3. Ready for use


Another point: I am testing version 1.5.0, but at the website, it still shows as the last version was two years ago:

https://plugins.jetbrains.com/plugin/13052-oracle-nosql-database-connector/versions

**Reply by dario-vega** [reply]
I already created an enhancement request - a better solution for the SDK pre-requirement. 

Otherwise, the current version is 1.5.1. We have two options to install the plugin:  
- Search the Oracle NoSQL Database Connector in the [JetBrains plug-in repository](https://docs.oracle.com/pls/topic/lookup?ctx=en/database/other-databases/nosql-database/24.1/admin&id=jetbrains_plugin_repo), and install it, or
- Download the IntelliJ plugin from [Oracle Technology Network](https://docs.oracle.com/pls/topic/lookup?ctx=en/database/other-databases/nosql-database/24.1/admin&id=intellij_plugin_download), and install the plugin from disk.

Unfortunately, the latest version is not currently available in the JetBrains plug-in repository. We are working on fixing it.

**Comment by otaviojava** [comment]
Please, could we enhance the documentation

The terminology is crucial when working with various tools and libraries. Misunderstanding these terms can lead to issues during implementation. 

#### Example: The SDK

Oracle's SDK documentation can be found at the following link:
[Oracle NoSQL Cloud SDK Documentation](https://docs.oracle.com/en/cloud/paas/nosql-cloud/uapic/)

Initially, I assumed that the SDK referred to the JAR file used in a Maven project. However, I encountered an issue when I attempted to set the SDK path to the Maven repository (`.m2/repository`). This approach did not work.


The mistake arose from a misunderstanding of the SDK installation instructions:

> "Complete path to the directory where you extracted the Oracle NoSQL Java SDK. For example..."

For further clarification, refer to the detailed connection instructions:
[Connecting to Oracle NoSQL Database with IntelliJ](https://docs.oracle.com/en/database/other-databases/nosql-database/22.3/admin/connecting-oracle-nosqldatabase-intellij.html)

**Reply by dario-vega** [reply]
I agree; in the following sentence, the link provided is not good; it is incredibly confusing.

> Download and extract Oracle NoSQL Java SDK. See [About Oracle NoSQL Database SDK Drivers](https://docs.oracle.com/pls/topic/lookup?ctx=en/database/other-databases/nosql-database/24.1/admin&id=UAPIC-GUID-6DB8FA3D-1E31-49E6-A54E-908F187968A2).

You are supposed to
1. download the latest driver/SDK from GitHub - [https://github.com/oracle/nosql-java-sdk/releases/latest](https://github.com/oracle/nosql-java-sdk/releases/latest)
2. extract (unzip) in a local repository
3. provide the path where you extracted the Oracle NoSQL Java SDK

It will be reviewed entirely in the next version. We want to provide a better solution for the Java SDK pre-requirement. We do not want to ask the end user to download the zip file from GitHub; instead, we want to execute it programmatically.

**Reply by dario-vega** [reply, marked as answer]
@otaviojava let me know if it is now clear

- https://docs.oracle.com/en/database/other-databases/nosql-database/24.1/admin/setting-intellij-plug.html
- https://docs.oracle.com/en/cloud/paas/nosql-cloud/yooud/index.html#GUID-BC240432-CE20-448A-8A8B-36589A1E6ACB

Again, we want to review it in the next version to provide a better experience - no need to download the driver manually

**Reply by otaviojava** [reply]
Way better @dario-vega 
I will prepare a tutorial how to use it and publish at my social media next week.
thanks.
